### PR TITLE
Add review websocket

### PR DESCRIPTION
**Pull request by TimothyGillespie**

### CODE REVIEW

File: src/main/kotlin/io/beatmaps/ws/server.kt [hunk 1]
```diff
@@ -72,6 +72,18 @@ fun Application.ws() {
...
+                queueBind("ws.reviewStream.created", "beatmaps", "reviews.*.created")
+
+                queueDeclare("ws.reviewStream.updated", true, false, false, genericQueueConfig)
+                queueBind("ws.reviewStream.updated", "beatmaps", "reviews.*.updated")
+
+                queueDeclare("ws.reviewStream.deleted", true, false, false, genericQueueConfig)
+                queueBind("ws.reviewStream.deleted", "beatmaps", "reviews.*.deleted")
+
+                queueDeclare("ws.reviewStream.curated", true, false, false, genericQueueConfig)
+                queueBind("ws.reviewStream.curated", "beatmaps", "reviews.*.curated")
```
Comment (Top-Cat) [inline src/main/kotlin/io/beatmaps/ws/server.kt]: If you want consistent message order you probably don't want 4 queues.
Just bind `reviews.*`

Yes, you'll then have to unpick the type of message later (maps send extra data in the message itself but it would also be possible to get the original routingKey from `message.envelope.routingKey`)

File: src/main/kotlin/io/beatmaps/ws/routes/reviews.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                    Review.deletedAt.isNull()
+                } else {
+                    Op.TRUE
+                }
+            }
+            .firstOrNull()?.let { row ->
+                ReviewWebsocketDTO.wrapRow(row)
+            }
+    }?.let { summary ->
+        val wsMsg = inlineJackson.writeValueAsString(WebsocketMessage(messageType, summary))
```
Comment (Top-Cat) [inline src/main/kotlin/io/beatmaps/ws/routes/reviews.kt:55]: Beyond the scope of this but we should switch from jackson to kotlin serialisation, you've added `@Serializable` but currently that does nothing.

File: src/main/kotlin/io/beatmaps/ws/routes/reviews.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        fun wrapRow(row: ResultRow): ReviewWebsocketDTO {
+            return ReviewWebsocketDTO(
+                row[Review.userId].value,
+                row[Review.mapId].value,
+                row[Review.text],
+                row[Review.sentiment],
+                row[Review.createdAt].toKotlinInstant().toString(),
+                row[Review.updatedAt].toKotlinInstant().toString(),
+                row[Review.curatedAt]?.toKotlinInstant()?.toString(),
+                row[Review.deletedAt]?.toKotlinInstant()?.toString()
```
Comment (Top-Cat) [inline src/main/kotlin/io/beatmaps/ws/routes/reviews.kt]: Generally if the row has been deleted I'd rather not treat it as still existing. See map deleted where we just return the id. In this case it would have to be the userId and mapId but everything else unset. Maybe a seperate constructor for that?

Reviews especially can be un-deleted by making a new review so you have an edge case where the deleted update actually returns a review that isn't deleted at all.

**Review by Top-Cat — CHANGES_REQUESTED** [review]

**Comment by Top-Cat** [comment]
Also.... Ok to test

**Comment by Top-Cat** [comment]
Already did m8

File: src/main/kotlin/io/beatmaps/ws/routes/reviews.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            )
+        }
+    }
+}
+
+suspend fun retrieveAndSendReview(messageType: WebsocketMessageType, reviewCompositeId: ReviewUpdateInfo, holder: ChannelHolder, onlyNonDeleted: Boolean) {
+    transaction {
+        Review
```
Comment (Top-Cat) [inline src/main/kotlin/io/beatmaps/ws/routes/reviews.kt:47]: fwiw I wouldn't be against adding fields to ReviewUpdateInfo (I think we can get the full state pretty much everywhere it's published) and then you wouldn't have to hydrate the state here at all

File: src/main/kotlin/io/beatmaps/ws/server.kt (hunk 1 above)
Comment (TimothyGillespie) [reply src/main/kotlin/io/beatmaps/ws/server.kt]: I currently don't have access to the `message.envelope` object and copying the code from consumeAck to gain access, seems kinda dirty. But using `message.envelope.routingkey` seems preferable. I'm really not sure how to utilize it though without making a mess.

Adding a third parameter to `consumeAck` has crossed my mind, but that doesn't seem great either.

File: src/main/kotlin/io/beatmaps/ws/routes/reviews.kt (hunk 3 above)
Comment (TimothyGillespie) [reply src/main/kotlin/io/beatmaps/ws/routes/reviews.kt]: I see your point! From a websocket user perspective it's probably nice to see what the content has been - to get that he must have known what the review was prior, but at the same time we might not want to be that verbose about it.

I addressed it.

**Comment by TimothyGillespie** [comment]
This is still untested since I didn't set up the project to use the local state of the common package

**Comment by TimothyGillespie** [comment]
I think this should be good now.

**Review by Top-Cat — APPROVED** [review]
